### ISSUE TITLE
automatic reload when file changes externally

### ISSUE THREAD

**Issue by dead10ck** [helix-editor/helix#1125]
It would be great if it were detected when the file you are editing has been changed externally, and if possible, automatically reload the file.

I imagine it could work similarly to vim and kakoune, i.e. if there are no modifications that have yet to be written, just reload the file, and treat it as a single modification that can be undone. If there are modifications that haven't been written yet, prompt the user about what they would like to do:

* Ignore the changes, keep that's in the editor buffer, and let a write happen over the external change, whatever it was
* Reload the file anyway, essentially "adding" to the unsaved changes

**Comment by archseer** [comment]
See https://github.com/helix-editor/helix/pull/588

**Comment by jtrees** [comment]
Just some quick feedback:

I'm currently trying Helix as an alternative to Kakoune and the lack of file watching really hurts. I frequently have several windows open with Helix where I work on the same files. Since there is no server-client architecture, the buffers get out-of-sync and make this workflow virtually unusable.

I suppose the most viable workaround is to use splits in a single Helix instance, instead of letting my window manager handle the windows.

**Comment by ryanfowler** [comment]
👋 Hi @pascalkuthe! I see you self-assigned this issue earlier in the year, but I would love to help move this feature along if possible.

I've taken a look at https://github.com/helix-editor/helix/pull/2653, but it seems like that branch has stalled (and doesn't include reloading open buffers). I'd like to propose some functionality that could be broken up into multiple steps/PRs:

**File watching and reloading open buffers**
This first step (and likely also the largest) would involve adding "file watching" functionality to the editor. I've looked into both the [`notify`](https://docs.rs/notify/latest/notify/index.html) crate, in addition to using an external tool like [`watchman`](https://facebook.github.io/watchman/). I think that using `notify` is the better approach that doesn't make Helix reliant on external tooling. I think a reasonable first implementation would watch the files in the current working directory, and take a list of excludes (probably globs?) to cut down on noise. VSCode uses some [default excludes](https://code.visualstudio.com/docs/getstarted/settings#_default-settings) (`"files.watcherExclude"`) that seem sensible. Once file watching is in place, open buffers that haven't been modified can be reloaded on change.

**Watching individual files that are opened, but not in the CWD**
For files that are opened but not watched because they are outside of the CWD, they can be separately watched via individual requests to the [`Watcher`](https://docs.rs/notify/latest/notify/trait.Watcher.html).

**Support LSP didChangeWatchedFiles**
Add support for [`dynamic_registration`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#didChangeWatchedFilesClientCapabilities) with LSPs, and forward requested file changes as necessary. I think this should be quite straightforward once file watching is already in place. Changed files can be compared against the patterns that LSPs request, and sent if there are any matches.

**Optimizations for file watching**
Specifically for inotify on linux, the exclude globs should be taken into account when traversing the file system to cut down on the number of paths that need to be watched. However, this is _likely_ only a problem on a small number of very large repositories, or on less powerful hardware.

If there is already work underway on this topic, then feel free to dismiss the above. But if not, I'd be interested in people's thoughts and happy to start with the work. Thanks!

**Comment by pascalkuthe** [comment]
> 👋 Hi @pascalkuthe! I see you self-assigned this issue earlier in the year, but I would love to help move this feature along if possible.
> 
> I've taken a look at #2653, but it seems like that branch has stalled (and doesn't include reloading open buffers). I'd like to propose some functionality that could be broken up into multiple steps/PRs:
> 
> **File watching and reloading open buffers** This first step (and likely also the largest) would involve adding "file watching" functionality to the editor. I've looked into both the [`notify`](https://docs.rs/notify/latest/notify/index.html) crate, in addition to using an external tool like [`watchman`](https://facebook.github.io/watchman/). I think that using `notify` is the better approach that doesn't make Helix reliant on external tooling. I think a reasonable first implementation would watch the files in the current working directory, and take a list of excludes (probably globs?) to cut down on noise. VSCode uses some [default excludes](https://code.visualstudio.com/docs/getstarted/settings#_default-settings) (`"files.watcherExclude"`) that seem sensible. Once file watching is in place, open buffers that haven't been modified can be reloaded on change.
> 
> **Watching individual files that are opened, but not in the CWD** For files that are opened but not watched because they are outside of the CWD, they can be separately watched via individual requests to the [`Watcher`](https://docs.rs/notify/latest/notify/trait.Watcher.html).
> 
> **Support LSP didChangeWatchedFiles** Add support for [`dynamic_registration`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#didChangeWatchedFilesClientCapabilities) with LSPs, and forward requested file changes as necessary. I think this should be quite straightforward once file watching is already in place. Changed files can be compared against the patterns that LSPs request, and sent if there are any matches.
> 
> **Optimizations for file watching** Specifically for inotify on linux, the exclude globs should be taken into account when traversing the file system to cut down on the number of paths that need to be watched. However, this is _likely_ only a problem on a small number of very large repositories, or on less powerful hardware.
> 
> If there is already work underway on this topic, then feel free to dismiss the above. But if not, I'd be interested in people's thoughts and happy to start with the work. Thanks!

I agree that depending in an external binary like watchmen is not a good idea. However the notify crate is not just non-optimal its implementation is exactly what the kernel documentation tells you *not* to do. The API is also pretty limited and doesn't play that well with tokio.  So to solve this I wanted to implement my own file watcher. Haven't gotten around to that yet tough

**Comment by ryanfowler** [comment]
Ah okay, that's fair! 

I can create a separate issue for this, but until a file watcher is in place, would you be opposed to adding support for the LSP didChangeWatchedFiles and send events when relevant documents are written to, from within Helix? This would resolve a common pain-point I have of modifying metadata files (e.g. `Cargo.toml`) in Helix and having to restart the LSP to pull in the changes. That work could then be replaced when full file watching functionality is added sometime in the future.

**Comment by pascalkuthe** [comment]
That could be a reasonable interim solution to simply send events into a channel in our save implementation and then handle those events in a watcher tokio task. That task can then handle the LSP related stuff (and will already be nontrivial) and keeps the whole thing decoupled from the actual watcher/out of the main code. That does seem like a reasonable first step to me.

You can ignore registering "interest" for now I think (so passing the glob pattern to the actual watcher n the future). That's pretty complex and I will probably initially startout watching the entire workspace anyway (I would ignore gitignored file tough to avoid crawling build artifacts). That's a complex problem and requires more careful API design.

Btw for a really nice experience I would also send a filewatcher event when we reload inside helix so that if the file changes externally you can still let the LSP know with `:reload`

**Comment by deifactor** [comment]
For what it's worth, I'm running into this a lot where I save my files, run `cargo clippy --fix --allow-staged`, and then I have to remember to `:rla` before I continue to work on them. I definitely agree with @pascalkuthe that "reload open files" and "tell LSP when a file has changed" are separate issues.

Would it be okay to just get an initial draft of this functionality that uses `notify` and only implements the 'reload currently-open files if they changed on disk and are not dirty' part? I'd be up for implementing that.

**Comment by TheLostLambda** [comment]
This is one of the last few things I really feel I'm missing in Helix — would I be right in thinking the LSP side of this was sorted in #7665 ?

Does that mean what's missing now is just buffer reloading when files change on the disk?

Let me know if that's the case and if there is anything I can do to help out!

**Comment by rcorre** [comment]
I find that I mostly run into this because I suspend helix, run some CLI operations, then resume helix. Reloading changes upon resuming from suspension would fix 90% of this for me. Would that be simpler than implementing a file watcher?

**Comment by alxpettit** [comment]
@rcorre 

Just one person here, but I should probably mention that wouldn't satisfy my use case. I almost always am working simultaneously in an editor and in a terminal. The terminal side of things is often making changes to files (via formatting tools, fixers, sops editing, etc), after which point, I have to manually restart Helix every time, or call `:reload-all`. This would keep happening regardless of any 'save on suspend' feature.

**Comment by webdev23** [comment]
As a temporary workaround, we could separately watch for some file changes in a given folder, and send some keystrokes to a byobu/tmux session.

Of course, `helix` has to run into a `tmux` session for this to work. 

Using inotifywait: `sudo apt install inotify-tools`

In this simplest example, that command is running into a `byobu` split, the Helix editor is in the first split. It does reload the file effectively, when `helix` is in normal mode.

` while inotifywait -e modify ~/Desktop; do byobu send-keys -t 0 ":rla" Enter; done; `

Using `tmux` only, we have to pass in the session name and the split number. It allows to be ran in the background. The `-r` flag is for recursive.

`while inotifywait -r -e modify ~/Desktop; do tmux send-keys -t mytmuxsession:0 ":rla" Enter; done;`

It works, but of course, this feels very hacky, because it is.
Looking forward for such watcher capabilities built-in.

It could be nice that `helix` accepts a given POSIX signal to trigger a reload.

Links to help building such watcher script:
https://superuser.com/a/181543 
https://unix.stackexchange.com/a/409863

**Comment by alxpettit** [comment]
> It could be nice that helix accepts a given POSIX signal to trigger a reload.

Yeah, I was thinking the same thing. How does everyone feel about this?

**Comment by rcorre** [comment]
> At that stage you could simply check every X number of key presses for the active buffer if the source file has changed and if so notify the user about it

It's not the best experience to find out a few keypresses in that the file changed. I'd rather know right away.

> A POSIX signal is a nice addition but the practical use feels extremely limited and it also would not work on non-posix platforms.

Agreed, but it sounds like there's a lot of pushback against a file monitor, and this feels like a pretty lightweight solution that would satisfy a lot of folks. If we could detect SIGCONT, that would make me happy.

**Comment by deifactor** [comment]
Is there pushback against a file monitor? The proposed solutions such as "check every X keystrokes" or "just reload on SIGCONT" seem more like workarounds to the lack of a file watcher rather than the ideal solution, and have flaws of their own (the former leads to lost of pointless disk IO, the latter has no way to indicate *which* file to reload).

**Comment by HealsCodes** [comment]
> Is there pushback against a file monitor? The proposed solutions such as "check every X keystrokes" or "just reload on SIGCONT/some other signal" seem more like workarounds to the lack of a file watcher rather than the ideal solution, and have flaws of their own (the former leads to lost of pointless disk IO, the latter has no way to indicate *which* file to reload and also requires a very specific workflow).

If it's a monitor that runs inside helix that would be fine as it could limit the checks to when helix is active / the visible buffers. In that regard it would be a clean replacement to my 'x' keystrokes (which btw is next to no io overhead as all you do is query the metadata not the file contents). 

An external monitor is shooting canons at birds as it would likely monitor and signal helix for multiple files without the ability to detect if helix is presently active (vs. suspended with ctrl-z or simply tabbed out). Nor would it have any ability to know which buffer is active in helix, which then would cause unnecessary signalling for buffers the users doesn't have visible right now and which would need to be discarded. So for that one: yes, clear pus back from at least one person 🤚

**Comment by deifactor** [comment]
> (an integrated watcher in helix would also guarantee it is available on all platforms helix is on vs. the inotify suggested further up)

Well, you still need to implement the watcher, at which point you use inotify or equivalent unless you want to poll on disk.

Overall I do think using the OS's file-watcher system is the best approach here, since it's literally what they're there for and all the major operating systems (Linux, OSX, Windows, BSDs) have one.

**Comment by porridgewithraisins** [comment]
another reason this is required: the colors to the left marking git status don't update if I do git commit in another terminal

**Comment by TheLostLambda** [comment]
Checking in with @pascalkuthe who's down as an assignee — do you have a minimum set of requirements / functionality that you'd like to see for a first draft of this functionality? Are there any partial solutions that would be acceptable?

From looking through things, it seems the obvious place is something that loops into inotify and plays nice with Tokio (would a Linux-only first pass at things be acceptable, or would you like to support all OS's from the get-go?).

For that in-memory filesystem tree you mentioned, what sort of "queries" are you interested in there? Is this another thing that could be added later down the line?

If you have the time, I'd be lovely to get a few indicators as to what you'd see as an acceptable contribution, then some of us with spare time can gradually work our way towards that vision! Working with several Helix instances open in Zellij means that I often run into this limitation and I'm more than happy to try to make some progress on it :)

Thanks a ton!

**Comment by pascalkuthe** [comment]
The file watcher needs to be a seperate library that should be working well before any PR to helix is amde. It should essentially be a rust version of watchman (as a library instead of as an external process). I got started with ut but it's pretty complex to implement.

**Comment by OldManFroggy** [comment]
just run into this issue while building a lsp, and testing using Helix, kind of annoying to `:rl` (reload) all of the time.

I am late to the party, so sorry if this is also covered earlier or elsewhere.

but, is there a way to just add a `:watch` command that works on active buffer to watrch to see if that file changes, this way you can quick build (yeah, a little hacky) a buffer based watch routine in Helix. (or even simpler, on selecting a buffer check file chanegs, if active buffer just do on a simple timer that starts from a keypress in the buffer and tracks time since last keypress, kinda thang)

this solves issue of and problems with a watch every file solution, and just doing a `:watch` `:watch-off` is a lot less hassle.

the tmux / external tools ideas are not user friendly, and doing watchman like tooling and forcing watching across all buffers will not be gerat for those that do not want or need to see an active buffer update.

anyway just my 2c, after hitting `rl` too manyt times, and simply now using nvim in seperate session to watch the file for updates live... :(

**Comment by webdev23** [comment]
> just run into this issue while building a lsp, and testing using Helix, kind of annoying to `:rl` (reload) all of the time.
> 
> I am late to the party, so sorry if this is also covered earlier or elsewhere.
> 
> but, is there a way to just add a `:watch` command that works on active buffer to watrch to see if that file changes, this way you can quick build (yeah, a little hacky) a buffer based watch routine in Helix. (or even simpler, on selecting a buffer check file chanegs, if active buffer just do on a simple timer that starts from a keypress in the buffer and tracks time since last keypress, kinda thang)
> 
> this solves issue of and problems with a watch every file solution, and just doing a `:watch` `:watch-off` is a lot less hassle.
> 
> the tmux / external tools ideas are not user friendly, and doing watchman like tooling and forcing watching across all buffers will not be gerat for those that do not want or need to see an active buffer update.
> 
> anyway just my 2c, after hitting `rl` too manyt times, and simply now using nvim in seperate session to watch the file for updates live... :(

I had very much the same issue, could simply not use Helix without auto-reloading, so I combined this, it does include a file watcher:

https://github.com/webdev23/shelix

**Comment by OldManFroggy** [comment]
@webdev23 will take a look later tonight, thanks for the link -> I went oldschool for time being and am now watching local stdout log using `tail -f ...` works good enough for LSP debugging, must admit Helix is the easiest editor by far for adding a new LSP dont get me started on vscode lsp dev.

**Comment by VKondakoff** [comment]
>  I got started with ut but it's pretty complex to implement.

Please, don't give up trying to implement this! Auto-reloading the file, which was changed externally is a must-have feature for a modern (and even post-modern) editor. ;)

**Comment by HealsCodes** [comment]
Just one small correction: offering an (optional) prompt to reload would be modern.

Sometimes your don't want it to autoreload at all costs or at least want to be able to press U and undo the external changes ;)

Wouldn't be the first time helix saved a file that became a victim of an accidental delete or overwrite - both scenarios where auto reload without prompt would have been fatal.

**Comment by ilyagr** [comment]
> Just one small correction: offering an (optional) prompt to reload would be modern.

I would prefer a prompt if and only if the file was modified both inside and outside the editor. If I haven't modified the file in helix, and it changed on disk, this likely means a VCS checkout, and I don't want a prompt.

**Comment by YaLTeR** [comment]
Besides, you don't need a prompt to be able to undo across reloads.

**Comment by VKondakoff** [comment]
What about implementation like this: let's introduce an option `on-external-change` (what to do when file changes externally) with following parameters: 
- `ignore` (do nothing, if the file is not "dirty") 
- `reload` (default)
- `tail` (reload and scroll to the end of the file)
- `prompt`

If the file is "dirty" (has unsaved changes) the prompt to save, reload or ignore should be fired.

**Comment by alxpettit** [comment]
@VKondakoff Sounds great! It could even be configured based on regex matches to the file name or somesuch. :)

**Comment by ilyagr** [comment]
> The file watcher needs to be a seperate library that should be working well before any PR to helix is amde. It should essentially be a rust version of watchman (as a library instead of as an external process).

I have no experience here but, naively, https://github.com/watchexec/watchexec/tree/main/crates/lib might work.

**Comment by gagbo** [comment]
> The file watcher needs to be a seperate library that should be working well before any PR to helix is amde. It should essentially be a rust version of watchman (as a library instead of as an external process). I got started with ut but it's pretty complex to implement.

Just so there’s a trace of the reasoning here, what are the issues that make you ignore the [notify crate](https://github.com/notify-rs/notify)? When I started looking into wrapping inotify sys crate I saw this, and I’ve been wondering what the problem is ever since. The unreliability of `inotify` on a big number of watchers?

**Comment by kirawi** [comment]
See https://github.com/helix-editor/helix/issues/8073#issuecomment-1902666294 regarding that

**Comment by quantonganh** [comment]
> Just one person here, but I should probably mention that wouldn't satisfy my use case. I almost always am working simultaneously in an editor and in a terminal. The terminal side of things is often making changes to files (via formatting tools, fixers, sops editing, etc), after which point, I have to manually restart Helix every time, or call `:reload-all`. This would keep happening regardless of any 'save on suspend' feature.

> another reason this is required: the colors to the left marking git status don't update if I do git commit in another terminal

If you are using [WezTerm](https://wezfurlong.org/wezterm/index.html), with Helix at the top and a terminal at the bottom, please check out [my workaround](https://quantonganh.com/2023/07/25/auto-reload-helix.md).

The idea is to [emit an event](https://wezfurlong.org/wezterm/config/lua/wezterm/on.html) when switching back from terminal (bottom) to the Helix editor (top):

```lua
wezterm.on('reload-helix', function(window, pane)
  local top_process = basename(pane:get_foreground_process_name())
  if top_process == 'hx' then
    local bottom_pane = pane:tab():get_pane_direction('Down')
    if bottom_pane ~= nil then
      local bottom_process = basename(bottom_pane:get_foreground_process_name())
      wezterm.log_info('bottom process: ' .. bottom_process)
      if bottom_process == 'lazygit' or bottom_process == 'fish' then
        local action = wezterm.action.SendString(':reload-all\r')
        window:perform_action(action, pane);
      end
    end
  end
end)
```

```lua
  {
    key = '[',
    mods = 'CMD',
    action = act.Multiple {
      act.ActivatePaneDirection 'Up',
      act.EmitEvent 'reload-helix',
    }
  },
```

**Comment by GetPsyched** [comment]
Is this being worked on? Seems like there's implementation discussion going on but no linked PR.

However, I just want to state that I agree with both sides (auto-reload and no auto-reload), hence I would prefer if there was a config option for this and I'm not forced into losing my unsaved changes just because something ran in the background.

**Comment by Jackojc** [comment]
Would really benefit from this feature in my workflow. Often working in a codebase where I need to rebase and files may become stale (then I might accidentally overwrite something without realising). Kakoune prompts me to reload which I think is the best of both worlds.

**Comment by TheZoq2** [comment]
Modified @quantonganh's wezterm solution a bit to reload any time window focus is changed

```lua
-- https://wezfurlong.org/wezterm/config/lua/pane/get_foreground_process_name.html
-- Equivalent to POSIX basename(3)
-- Given "/foo/bar" returns "bar"
-- Given "c:\\foo\\bar" returns "bar"
function basename(s)
  return string.gsub(s, '(.*[/\\])(.*)', '%2')
end

wezterm.on('window-focus-changed', function(window, pane)
    if window:is_focused() then
        local top_process = basename(pane:get_foreground_process_name())
        print(top_process)
        if top_process == 'hx' then
            local action = wezterm.action.SendString(':rla\r')
            window:perform_action(action, pane);
        end
    end
end)
```

**Comment by 7flash** [comment]
Actually love current behavior of helix allowing me to keep stale versions of my file, as others mentioned, very useful for comparison in between branches. But also need this feature as an alternative to "tail -f" for watching log files.

**Comment by TheZoq2** [comment]
> Modified @quantonganh's wezterm solution a bit to reload any time window focus is changed
> 
> ```lua
> -- https://wezfurlong.org/wezterm/config/lua/pane/get_foreground_process_name.html
> -- Equivalent to POSIX basename(3)
> -- Given "/foo/bar" returns "bar"
> -- Given "c:\\foo\\bar" returns "bar"
> function basename(s)
>   return string.gsub(s, '(.*[/\\])(.*)', '%2')
> end
> 
> wezterm.on('window-focus-changed', function(window, pane)
>     if window:is_focused() then
>         local top_process = basename(pane:get_foreground_process_name())
>         print(top_process)
>         if top_process == 'hx' then
>             local action = wezterm.action.SendString(':rla\r')
>             window:perform_action(action, pane);
>         end
>     end
> end)
> ```

Oh, I just noticed a potential pitfall with this: `reload` discard changes instead of requiring `reload!`  for changed files, so this is not a great solution :/

**Comment by kirawi** [comment]
VSCode seems to use https://github.com/parcel-bundler/watcher

**Comment by p00ya** [comment]
I think a much simpler and specific solution could be implemented.  Namely, call `stat` for the Document in the focused window when it's revised and show a vim-style popup if the mtime has changed before save-time.  This is a relatively low performance overhead (only a single path being polled at a time, in response to user edits) and even that could be further reduced with a debounce interval.

Push notifications via a file watcher could later be integrated as an *optimization*; either using the notify crate, whatever Pascal is working on (if it's ever finished), or something else.

The stat-based solution might not be fit for wider purposes like having a full FS cache for the LSP, but it seems to me like #1125 deserves a short-term solution given the number of whacky solutions people are implementing already (sending signals to helix or even piping keystrokes to their terminal emulator 🤢).

I'm happy to implement the stat-based approach for myself.  I tried to touch base with Pascal on Matrix a couple of weeks ago to discuss a PR but got no response.

> VSCode seems to use https://github.com/parcel-bundler/watcher

An external process like that is disagreeable for the [same reasons as watchman](https://github.com/helix-editor/helix/issues/1125#issuecomment-1636824064) - perhaps more so since it's Node bloatware vs a C++ binary.

While I'm here, it's worth discussing Pascal's reasons for objecting to the [notify crate](https://docs.rs/notify/latest/notify/index.html):

> The notify crate is the current goto solution in rust but it's too unreliable in my opinion. There is no automatic way to recover from dropped events (which can occur frequently, articels on the linux kernel mailing list explaining how to use inotify specically recommend not to use the architecture notify is using). Instead a client would have to restart notify in those situations which is very slow and makes downstream logic complicated.

This is an intrinsic problem with using a best-effort system API like inotify - the notify crate does at least expose `Event::needs_rescan`, and yes it requires good-old `readdir`/`stat` and making the inotify syscall again, and you probably want to do that periodically anyway.  Whatever performance penalty all that involves, you'd incur it no matter what you're wrapping the syscall up in - the best you can do is mitigate it by putting the syscall in a background thread and using epoll to multiplex the I/O (e.g. with tokio).

> There is also no way to query the current fs state (which a filewatcher needs to keep anyway to work properly so maintaining that downstream is both inefficient and quite cumbersome.

Just use `readdir`/`stat` (via tokio) to query the actual FS (which modern kernels will cache already).  For the purposes of #1125, the use of kernel APIs like inotify/fsevent is just an optimization that can save some of these other calls, not a replacement for them.

From Pascal's message on matrix:
> System threads are not optimized that way and just spawn 100 threads is going to cause a ton of overhead. Each thread has roughly the same priority so if we spawn 100 notify watchers they get more priority they take significant processing time away from tokio/the rest of the editor. Therefore notify offers an API to easily watch many different files with the same watcher (thread) however you then need a seperate tokio tasks that recives all files changed events and forwards them. That means you need to build some kind of registry to track which channel is interested in which is not trivial to do efficently.

I agree what the notify crate is doing is inferior to integrating with a tokio reactor, but the thread starvation concerns would be pretty minor for #1125 (i.e. there only needs to be a single inotify fd for all open Documents; you can have multiple watches on the one queue/thread).

Also, FWIW, a relatively small change to the notify crate would let it use tokio to wait on the inotify FD, rather than it spawning new threads.

Another issue that Pascal pointed out from reddit:
> notify tires to solve two very hard problems: papering over abysmal syscalls provided by various OSes, and implementing a robust model which prevents losing notifications. On the second point, I’ve found it to be insufficient for building robust functionality. Watching fs really requires defining watchman-style clock ids

Using a watchman-style serial clockid is only really relevant if you want to sync cached metadata, which I think is massively over-engineering the solution to #1125 compared to the `stat` approach I'm suggesting.  Helix is already naive/racy in its `Document::save_impl` implementation, which just compares the Document's mtime to the FS before writing.  Helix doesn't need to know about the order of events, just that the file changed externally.

**Comment by kirawi** [comment]
> An external process like that is disagreeable for the https://github.com/helix-editor/helix/issues/1125#issuecomment-1636824064 - perhaps more so since it's Node bloatware vs a C++ binary.

What I proposed is to rewrite it in Rust if a POC using bindings to the C++ library (the JavaScript is only for the WASM wrapper) pans out.

**Comment by p00ya** [comment]
I have a proof-of-concept for the polling approach in this branch: https://github.com/helix-editor/helix/compare/master...p00ya:helix:file-changed

I haven't implemented a menu/picker yet; it just indicates an external modification via `[!]` in the status line.  I've mainly been testing for correctness and performance, which seem okay so far (with a 2s debounce).  Not quite ready for a PR but early feedback welcome.

> What I proposed is to rewrite it in Rust if a POC using bindings to the C++ library (the JavaScript is only for the WASM wrapper) pans out.

Ah, I see.  That doesn't sound too bad.

**Comment by pirj** [comment]
Just for the reference, how this is done in [neo]vim:
```
W12: Warning: File "todo/weekly.md" has changed and the buffer was changed in Vim as well
```
when there's a conflicting unsaved change in the editor, while the file was synced from my phone.

**Comment by AnthonyMichaelTDM** [comment]
I've used the notify crate for similar behavior in my other projects and it worked quite well for me

https://github.com/AnthonyMichaelTDM/mecomp/blob/main/daemon/src/dynamic_updates.rs

**Comment by ezynda3** [comment]
I see this basic feature still doesn't exist...it's one of the reasons I went back to vim...

**Comment by pirj** [comment]
There is no consensus as to how the editor should behave when a file is externally updated.
What do you think of a configuration option:
```toml
[editor.file-changed-externally]
behaviour = ... # can be: nothing, reload-unless-edited, warn
```

Helix incredibly sane and intuitive defaults. But this one is what divides people. Are there any downsides to be able to configure this?

**Comment by dpc** [comment]
Probably want to split it into two valures: "clean buffer" and "modified buffer", each can be one of: "ignore", "ask", "reload"

**Comment by AnthonyMichaelTDM** [comment]
That makes perfect sense to me

**Comment by webdev23** [comment]
Can't believe that this issue is still there after years, while critically important.

You are still trying to figure out what fancy tokio library to use, while a single line CRC check could be handled by a 2bit eeprom coasting 0.0004ct during 30 years.

Scuse me, but this starts to be too opinionated. 

This feature is a major constraint for anything programming. It is mandatory now.

**Comment by porridgewithraisins** [comment]
Someone needs to add this thread to the wikipedia page for bikeshedding as the _primary_ example. Rust just exacerbates the effect as well.

**Comment by Rudxain** [comment]
> often making changes to files (via formatting tools, fixers, sops editing, etc), after which point, I have to manually restart Helix every time, or call `:reload-all`

I want to add that, [for some repos](https://github.com/rust-lang/rust-analyzer/issues/19402#issuecomment-3048557880), I'm no longer using Rust-Analyzer, so I can't use its auto-format feature, I have to manually `:! cargo fmt` then `:reload` (or config `rustfmt` as a formatter) to achieve the same thing.

Other users would have this problem with some other lang LSP.

> git commit the external change; save the buffer; see the conflict and resolve manually.

I've just tested `git stash` and, surprisingly, it can detect a "merge conflict" even if all of the changes are uncommitted! Instead of `commit`, do `stash`, make the changes in-editor, `:write`, `git add` the files you've edited (Even if you skip this step, Git won't overwrite conflicting files, but it won't include commit markers. The `add` is only necessary for markers to appear), then `git stash pop`. After this, the usual "merge conflict resolution mode" will be triggered

**Comment by g-gundam** [comment]
# Extend Helix with Scheme

A solution for this exists if you have Helix+Steel installed.

- https://github.com/mattwparas/helix/blob/steel-event-system/STEEL.md (Build instructions for Helix+Steel)
- https://github.com/helix-editor/helix/pull/8675
- https://github.com/mattwparas/helix-file-watcher (Automatically reload buffers that were externally modified.)

## Installation

```bash
forge pkg install --git https://github.com/mattwparas/helix-file-watcher.git
```

## Configuration

Add this to `~/.config/helix/init.scm`.

```scheme
(require "helix-file-watcher/file-watcher.scm")
(spawn-watcher)
```

**Comment by andreslucena** [comment]
> Can't believe that this issue is still there after years, while critically important.
> 
> (...)
> 
> This feature is a major constraint for anything programming. It is mandatory now.
> 

This is an open source project. It seems there''s some consensus on how to solve this. Feel free to open a PR instead of complaining in issues. 

For me, doing `:reload-all` whenever I change branches is a valid workaround and is not a deal breaker. If it is a problem for you, please just use any of the great alternatives to Helix (NeoVim, Kakoune, etc).

**Comment by kirawi** [comment]
https://github.com/pascalkuthe/filesentry

**Comment by HealsCodes** [comment]
> https://github.com/pascalkuthe/filesentry

is not yet viable as cross-platform support for all the systems Helix runs on is required and filesentry currently is linux only with macOS and windows "planned"

**Comment by webdev23** [comment]
> > Can't believe that this issue is still there after years, while critically important.
> > (...)
> > This feature is a major constraint for anything programming. It is mandatory now.
> 
> This is an open source project. It seems there''s some consensus on how to solve this. Feel free to open a PR instead of complaining in issues.
> 
> For me, doing `:reload-all` whenever I change branches is a valid workaround and is not a deal breaker. If it is a problem for you, please just use any of the great alternatives to Helix (NeoVim, Kakoune, etc).

There is many PR already that could improve this issue, but them are never rolled in, such as https://github.com/helix-editor/helix/pull/5911 ,  https://github.com/helix-editor/helix/pull/5983 , or https://github.com/helix-editor/helix/pull/13896 and so many others.

Now, if you want some staging concepts, again, CRC is the cheapest computation we could find, because it's exactly made to be called in redundancy, as its name. 

https://en.wikipedia.org/wiki/Cyclic_redundancy_check

In Rust:
https://rosettacode.org/wiki/CRC-32#Rust

---

Edit: I am loosing it at "please just use any of the great alternatives to Helix (NeoVim, Kakoune, etc)". 

This is opinion, Sir.  You either provide tools or opinions, but not both.

**Comment by deifactor** [comment]
@webdev23 I'm not sure why CRC is relevant here. If you have enough open files that the cost of comparing `in_memory == on_disk` vs `crc32(in_memory) == crc32(on_disk)` is expensive, you have enough open files that you shouldn't be comparing the contents *anyway* and should just use the OS's native 'notify me when this file changed' functionality, because it's the disk access that's the expensive part. (And that's ignoring the fact that comparing the contents lets you exit early if there's a change near the start, whereas the crc always requires reading the entire file.)

I'm definitely frustrated as well; this is probably the single biggest annoyance I have with Helix right now.

**Comment by Jackojc** [comment]
I don't think it's helpful for anyone to say "use another editor". I think this is a pretty reasonable feature request.

**Comment by kirawi** [comment]
> > https://github.com/pascalkuthe/filesentry
> 
> is not yet viable as cross-platform support for all the systems Helix runs on is required and filesentry currently is linux only with macOS and windows "planned"

From what I understood, they are looking at first merging file watching for Linux, and will add the MacOS and Windows backends later on.

**Comment by WizardUli** [comment]
> There is many PR already that could improve this issue, but them are never rolled in, such as [#5911](https://github.com/helix-editor/helix/pull/5911) , [#5983](https://github.com/helix-editor/helix/pull/5983) , or [#13896](https://github.com/helix-editor/helix/pull/13896) and so many others.

Btw. specifically #13896 seems to be just a latest of a long chain of PRs (not just issues or discussions) trying to solve utterly basic use-cases of integrating Helix with other tools. Think which should not even be called a feature but a missing core functionality. It's the same story as with this autoreload.

You can't say people don't try to contribute.

**Comment by porridgewithraisins** [comment]
Why doesn't doing what vim does suffice?

set autoread

And every `updatetime` interval, also everytime you focus vim, it loops through open buffers and compares the mtime of each backing file to the timestamp that file was last r/w from within vim.

Why can't helix do this?

**Comment by AnthonyMichaelTDM** [comment]
> Why doesn't doing what vim does suffice?
> 
> set autoread
> 
> And every `updatetime` interval, also everytime you focus vim, it loops through open buffers and compares the mtime of each backing file to the timestamp that file was last r/w from within vim.
> 
> Why can't helix do this?

That's a good idea, I'm going to throw together a prototype

**Comment by webdev23** [comment]
> [@webdev23](https://github.com/webdev23) I'm not sure why CRC is relevant here. If you have enough open files that the cost of comparing `in_memory == on_disk` vs `crc32(in_memory) == crc32(on_disk)` is expensive, you have enough open files that you shouldn't be comparing the contents _anyway_ and should just use the OS's native 'notify me when this file changed' functionality, because it's the disk access that's the expensive part. (And that's ignoring the fact that comparing the contents lets you exit early if there's a change near the start, whereas the crc always requires reading the entire file.)
> 
> I'm definitely frustrated as well; this is probably the single biggest annoyance I have with Helix right now.

You must not worry. We live in an era where each key type is sent to _node.js_ LSPs the process tree is incredible.

Otherwise, yes, i left another approach here already about 2 years ago, about `inotifywait` or it's counterpart `libnotify`, that is a Linux and mac thing, highly optimized.

Under the hood however, still, these systems are using CRC, as are anti-viruses in Windows, perpetually scanning the file tree and computing a Merkel tree of CRC hashes. It's made to be cheap as possible. Any hash system would do of course, and they are numerous. We could use sha256 it would be no harm, as our computer are highly optimized at the hardware level for that.

**Comment by deifactor** [comment]
> Under the hood however, still, these systems are using CRC

If that was the case, then inotify couldn't support the ability to notify when a file is opened, read, or closed, which [it does](https://man7.org/linux/man-pages/man7/inotify.7.html). Since filesystem access has to go through the kernel, the kernel can provide an API to let watchers know when a file is accessed without actually looking at the contents.

**Comment by webdev23** [comment]
Yeah whatever. Just implement auto reload, with whatever tech flavor you like. 

Thank you.

**Comment by webdev23** [comment]
> > Under the hood however, still, these systems are using CRC
> 
> If that was the case, then inotify couldn't support the ability to notify when a file is opened, read, or closed, which [it does](https://man7.org/linux/man-pages/man7/inotify.7.html). Since filesystem access has to go through the kernel, the kernel can provide an API to let watchers know when a file is accessed without actually looking at the contents.

In the source of `libnotify`, we could observe the usage of GHashTable .

```
 priv->action_map = g_hash_table_new_full (g_str_hash,  g_str_equal, ....

```

https://gitlab.gnome.org/search?search=hash&nav_source=navbar&project_id=1798&group_id=8&search_code=true&repository_ref=master

https://docs.gtk.org/glib/struct.HashTable.html


Guys, don't panic, but implementing auto reload could be done in one hour in a hacky night, but if after two+ years no solutions were found, it is either skill issues, opinions or both.

Nobody wants to use a program made of skill issues, tech debt, or opinions. It's my point of view, but I am not alone. 

Now, be nice, and focus on the basics. We spent time, we need a reliable Helix. Do not deceive us.

**Comment by jmagaram** [comment]
Adding a critical use case. I use Claude Code to modify my code. These AI tools are becoming more and more essential. I want to see the changes in the open buffer and not have to remember to type rla or worse to make changes and get errors when saving.

**Comment by dpc** [comment]
I'll take a risk of writing another comment over just adding 👍 to one above, just to underline the point.

Aider, and even more so Claude Code fundamentally changed how I approach coding, and thus my relation with the code editor. I don't hope/care that AI tools become better "soon" or anything. As they are right now, they are make typing in large chunks of daily code needs completely unnecessary, obsolete and just uneconomical. I am also not some AI-hypists, just pragmatic, evaluating tools as they are.

For some time time now, rarely do I type things myself directly (still happens, especially for tricky but small code-wise issues), and the usage of Helix shifted to being more about viewing, reading and small-scope editing code increasingly outputted by an LLM, which Helix is actually well suited for.

But this particular issue was a minor nuisance in the past, is becoming increasingly important, even though personally I already got into a habit of issuing `:reload-all` all the time. I just feel the lack in ergonomics of it, and it's bothering me more.

**Comment by zygotron** [comment]
@dpc My solution: I run lazygit, watch diffs live as {claude,open}code modify files, hit e in lazygit for fast snipe edits in hx of modifications if necessary. There are no constantly open editor windows at all, and it makes git the locus of control. (in lazygit config set refreshInterval: 1, and editPreset: "helix", and promptToReturnFromSubprocess: false, and you're good to go with a smooth as butter workflow, imo). I deliberated over if this was topical, in one sense it is: externalize tools, one tool for the job, the helix way is the unix way.

**Comment by sibouras** [comment]
so you load the lsp everytime you want to make a quick change, besides not everyone uses lazygit and claude and reloading a file doesn't make it not unixy in any way

**Comment by zygotron** [comment]
@sibouras #312 would definitely make this kind of thing a lot better. no not everyone uses lazygit, but that wasn't really the point, you could just `$ watch git status` i guess, and i'm all for helix getting file reloading. My point was having git at the center of that kind of workflow is actually not such a bad idea, and there are existing tools to facilitate that right now.

**Comment by zygotron** [comment]
As an aside #312 could also facilitate file reloading, especially if you could pipe commands to helix. Then you could have fine grained control over which files and when they were reloaded, and have it under agent or other tool control as well, e.g. agent->mcp->pipe->helix

**Comment by 7flash** [comment]
Currently, I'm using this command to open file in hx from nearby split tabs (either from claude, lazygit, or yazi)

```
kitty @ send-text --match title:hx :open /file/path \\r
```

**Comment by quantonganh** [comment]
Another workaround if you are using [WezTerm](https://wezterm.org/index.html):
- write a simple wrapper that watch for the files changes, and send the `:reload-all` command to the current pane

```sh
#!/usr/bin/env bash

cleanup() {
  kill $watchexec_pid 2>/dev/null
}
trap cleanup EXIT

watchexec --postpone --exts rs,go,java,md -- hx-reload &
watchexec_pid=$!

hx "$@"
```

- `hx-reload` will be something like this:

```sh
#!/usr/bin/env bash

hx_pane_id=$WEZTERM_PANE

# Send ESC first to make sure we're in normal mode
wezterm cli send-text --pane-id "$hx_pane_id" --no-paste -- "\x1b"

# Small delay before sending :reload
sleep 0.05

wezterm cli send-text --pane-id $hx_pane_id --no-paste -- ":"
wezterm cli send-text --pane-id $hx_pane_id -- "reload-all"
printf "\r" | wezterm cli send-text --pane-id $hx_pane_id --no-paste
```

Tested with `lazygit`, `opencode`, ... and it's working fine.

**Comment by jokeyrhyme** [comment]
@quantonganh it might even be fine to not trigger upon file change, but to trigger every time focus is given to the Wezterm pane that is running helix? sort of paying for simplicity with a bit worse performance 🤷

**Comment by Zentropivity** [comment]
Another workaround: a language server that watches opened files and applies changes, this exists now: [fs_watcher_lsp](https://crates.io/crates/fs_watcher_lsp)

**Comment by webdev23** [comment]
> Another workaround: a language server that watches opened files and applies changes, this exists now: [fs_watcher_lsp](https://crates.io/crates/fs_watcher_lsp)

Oh sure, that would work, but to trigger the refresh an API is needed.
Of course, some made it already: https://github.com/helix-editor/helix/pull/13896

But some others are just too stubborn and incompetent to push the green button: https://github.com/helix-editor/helix/issues/405#issuecomment-3166065059

I am not saying we could do better. I am saying we are all taken in hostage by stupidity.

**Comment by noor-tg** [comment]
so is there auto-reload feature in helix ? or any other workaround ? beside the `wezterm`

**Comment by quantonganh** [comment]
> so is there auto-reload feature in helix ? or any other workaround ? beside the `wezterm`

https://github.com/helix-editor/helix/pull/14544

**Comment by eerii** [comment]
Here is another workaround for those using [niri](https://github.com/niri-wm/niri). It works similarly to the wezterm solution, reloading each time you focus the window, but it uses `wtype` to send the keybind. Not very optimized but it seems to work quite fluently.

It should be terminal independent, and I'm sure it can be adapted to other wms like sway. Written in fish but the idea is applicable to bash and others.

```fish
#!/usr/bin/env fish

set -l last_id ""

niri msg --json event-stream | while read -l line
    if string match -r '"WindowFocusChanged":\s*\{\s*"id":\s*(?<current_id>\d+)\s*\}' $line >/dev/null
        if test "$current_id" = "$last_id"
            continue
        end
        set last_id $current_id

        set -l window (niri msg --json focused-window)
        string match -r '"app_id":\s*"(?<app_id>[^"]+)"' $window >/dev/null
        string match -r '"title":\s*"(?<title>[^"]+)"' $window >/dev/null

        if test "$app_id" = foot; and string match -r '^hx' $title >/dev/null
            sleep 0.05
            wtype -M ctrl r -m ctrl
        end
    end
end
```

```toml
[keys.normal]
"C-r" = ":reload-all"

[editor.auto-save]
focus-lost = true # Works nice with save on focus lost :)
```
